### PR TITLE
Emit EnC info for methods with no sequence points

### PR DESCRIPTION
**Pull request by tmat**
Fixes https://github.com/dotnet/roslyn/issues/55381

### CODE REVIEW

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs [hunk 1]
```diff
@@ -73,78 +73,81 @@ public void SerializeDebugInfo(IMethodBody methodBody, StandaloneSignatureHandle
...
+            // bug #1067635 is fixed and available in SDK.
+            bool suppressNewCustomDebugInfo = compilationOptions.OutputKind == OutputKind.WindowsRuntimeMetadata;
+
+            bool emitDynamicAndTupleInfo = emitAllDebugInfo && !suppressNewCustomDebugInfo;
+
+            // Emit EnC info for all methods even if they do not have sequence points.
+            // The information facilitates reusing lambdas and closures. The reusal is important for runtimes that can't add new ones (e.g. Mono).
+            bool emitEncInfo = compilationOptions.EnableEditAndContinue && _metadataWriter.IsFullMetadata && !suppressNewCustomDebugInfo;
+
+            byte[] blob = customDebugInfoWriter.SerializeMethodDebugInfo(Context, methodBody, methodHandle, emitStateMachineInfo: emitAllDebugInfo, emitEncInfo, emitDynamicAndTupleInfo, out bool emitExternNamespaces);
```
Comment (davidwengier) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs:91]: If I'm reading this right, this changes so the custom debug info is never written for WindowsRuntimeMetadata? Seems like the previous code would write some of it (state machines for example)

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs [hunk 2]
```diff
@@ -73,78 +73,81 @@ public void SerializeDebugInfo(IMethodBody methodBody, StandaloneSignatureHandle
...
             bool isKickoffMethod = methodBody.StateMachineTypeName != null;
-            bool emitDebugInfo = isKickoffMethod || !methodBody.SequencePoints.IsEmpty ||
+            bool emitAllDebugInfo = isKickoffMethod || !methodBody.SequencePoints.IsEmpty ||
                 methodBody.MethodDefinition == (Context.Module.DebugEntryPoint ?? Context.Module.PEEntryPoint);
 
-            if (!emitDebugInfo)
+            var compilationOptions = Context.Module.CommonCompilation.Options;
+
+            // We need to avoid emitting CDI DynamicLocals = 5 and EditAndContinueLocalSlotMap = 6 for files processed by WinMDExp until 
+            // bug #1067635 is fixed and available in SDK.
```
Comment (davidwengier) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs:82]: This bug is never getting fixed, right? 😁

**Review by davidwengier — APPROVED** [review]

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs (hunk 2 above)
Comment (tmat) [reply src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs:82]: Probably not.

**Review by Wincell10 — APPROVED** [review]

**Review by Wincell10 — APPROVED** [review]

**Review by Wincell10 — APPROVED** [review]

**Comment by lambdageek** [comment]
Can this be merged?

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler PTAL (@cston)

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs [hunk 3]
```diff
@@ -73,78 +73,81 @@ public void SerializeDebugInfo(IMethodBody methodBody, StandaloneSignatureHandle
...
+            bool suppressNewCustomDebugInfo = compilationOptions.OutputKind == OutputKind.WindowsRuntimeMetadata;
+
+            bool emitDynamicAndTupleInfo = emitAllDebugInfo && !suppressNewCustomDebugInfo;
+
+            // Emit EnC info for all methods even if they do not have sequence points.
+            // The information facilitates reusing lambdas and closures. The reusal is important for runtimes that can't add new ones (e.g. Mono).
```
Comment (cston) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs:88]: >reusal [](http://example.com/codeflow?start=77&length=6)

"reuse" perhaps?

Comment (cston) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs:88]: >ones [](http://example.com/codeflow?start=129&length=4)

"methods"?

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs [hunk 4]
```diff
@@ -53,98 +53,95 @@ private void SerializeMethodDebugInfo(IMethodBody bodyOpt, int methodRid, int ag
...
-                    SerializeAsyncMethodSteppingInfo(asyncInfo, methodHandle, aggregateMethodRid);
-                }
+                SerializeStateMachineLocalScopes(bodyOpt, methodHandle);
             }
 
-            SerializeStateMachineLocalScopes(bodyOpt, methodHandle);
-
-            // delta doesn't need this information - we use information recorded by previous generation emit
+            // Emit EnC info for all methods even if they do not have sequence points.
+            // The information facilitates reusing lambdas and closures. The reusal is important for runtimes that can't add new ones (e.g. Mono).
```
Comment (cston) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs:143]: >reusal is important for runtimes that can't add new ones [](http://example.com/codeflow?start=77&length=56)

Same comments here as PdbWriter.cs

**Review by cston — APPROVED** [review]
Minor comments only.

**Review by jaredpar — APPROVED** [review]

**Pull request merged**